core/drivers/clk/clk.c

```diff
@@ -14,82 +15,154 @@
+	if (!clk_elt)
 		return NULL;
 
-	clk->num_parents = parent_count;
+	clk_elt->num_parents = parent_count;
 	for (parent = 0; parent < parent_count; parent++)
-		clk->parents[parent] = parent_clks[parent];
+		clk_elt->parents[parent] = parent_clks[parent];
 
-	clk->name = name;
-	clk->ops = ops;
-	refcount_set(&clk->enabled_count, 0);
+	clk_elt->name = name;
+	clk_init_instance(&clk_elt->clk, ops);
+
+	return &clk_elt->clk;
+}
+
+struct clk *clk_lw_alloc(const struct clk_ops *ops, size_t count)
+{
+	struct clk *clk = NULL;
+	size_t n = 0;
+
+	assert(ops->id == CLK_OPS_LIGHTWEIGHT);
+
+	clk = calloc(count, sizeof(*clk));
+	if (!clk)
+		return NULL;
+
+	for (n = 0; n < count; n++)
+		clk_init_instance(clk + n, ops);
 
 	return clk;
 }
 
 void clk_free(struct clk *clk)
 {
-	free(clk);
+	if (clk && is_clk_elt(clk))
+		free(clk_to_clk_elt(clk));
+	else
+		free(clk);
 }
 
 static bool __maybe_unused clk_check(struct clk *clk)
 {
-	if (!clk->ops)
+	if (!clk->ops) {
+		DMSG("Missing ops reference on clock");
 		return false;
+	}
 
-	if (clk->ops->set_parent && !clk->ops->get_parent)
-		return false;
+	if (is_clk_elt(clk)) {
+		struct clk_elt *clk_elt = clk_to_clk_elt(clk);
 
-	if (clk->num_parents > 1 && !clk->ops->get_parent)
-		return false;
+		if (clk->ops->set_parent && !clk->ops->get_parent) {
+			DMSG("set_parent but no get_parent ops on clock %s",
+			     clk_get_name(clk));
+			return false;
+		}
+
+		if (clk_elt->num_parents > 1 && !clk->ops->get_parent) {
+			DMSG("Several parents but no get_parent op on clock %s",
+			     clk_get_name(clk));
+			return false;
+		}
+
+		return true;
+	}
+
+	if (is_clk_lw(clk)) {
+		if (clk->ops->set_parent || clk->ops->get_parent) {
+			DMSG("Unpexpected parent clock ops on clock %s",
+			     clk_get_name(clk));
+			return false;
+		}
+
+		return true;
+	}
 
-	return true;
+	return false;
 }
 
-static void clk_compute_rate_no_lock(struct clk *clk)
+static void clk_elt_cache_rate_no_lock(struct clk *clk)
 {
 	unsigned long parent_rate = 0;
+	struct clk_elt *clk_elt = NULL;
 
-	if (clk->parent)
-		parent_rate = clk->parent->rate;
+	if (!is_clk_elt(clk))
+		return;
+
+	clk_elt = clk_to_clk_elt(clk);
+	if (clk_elt->parent)
+		parent_rate = clk_get_rate(clk_elt->parent);
 
-	if (clk->ops->get_rate)
-		clk->rate = clk->ops->get_rate(clk, parent_rate);
+	if (clk->ops->compute_rate)
+		clk_elt->rate = clk->ops->compute_rate(clk, parent_rate);
 	else
-		clk->rate = parent_rate;
+		clk_elt->rate = parent_rate;
 }
 
 struct clk *clk_get_parent_by_index(struct clk *clk, size_t pidx)
 {
-	if (pidx >= clk->num_parents)
+	if (!is_clk_elt(clk))
 		return NULL;
 
-	return clk->parents[pidx];
+	if (pidx >= clk_to_clk_elt(clk)->num_parents)
+		return NULL;
+
+	return clk_to_clk_elt(clk)->parents[pidx];
 }
 
 static void clk_init_parent(struct clk *clk)
 {
+	struct clk_elt *clk_elt = clk_to_clk_elt(clk);
 	size_t pidx = 0;
 
-	switch (clk->num_parents) {
+	if (!is_clk_elt(clk))
+		return;
+
+	clk_elt = clk_to_clk_elt(clk);
```

Comment:
`clk_elt` is already assigned ligne 148.
I guess the first cast is done a bit too early before checking the clock is indeed a clk_elt. You should remove the first one.

Reply:
thanks, i should have removed the one at line 148.